packages/util-dynamodb/src/convertToAttr.spec.ts

```diff
@@ -0,0 +1,220 @@
+      });
+    });
+  });
+
+  describe("list", () => {
+    const arr = [...Array(4).keys()];
+    const uint8Arr = new Uint32Array(arr);
+    const biguintArr = new BigUint64Array(arr.map(BigInt));
+    [
+      [
+        [null, false],
+        [{ NULL: true }, { BOOL: false }],
+      ],
+      [
+        [1.01, BigInt(1), "one"],
+        [{ N: "1.01" }, { N: "1" }, { S: "one" }],
+      ],
+      [
+        [uint8Arr, biguintArr],
+        [{ B: uint8Arr }, { B: biguintArr }],
+      ],
+    ].forEach(([input, output]) => {
+      it(`testing list: ${input}`, () => {
+        // @ts-ignore
+        expect(convertToAttr(input)).toEqual({ L: output });
+      });
+    });
+  });
+
+  describe("set", () => {
+    it("number set", () => {
+      const set = new Set([1, 2, 3]);
+      expect(convertToAttr(set)).toEqual({ NS: Array.from(set).map((num) => num.toString()) });
+    });
+
+    it("bigint set", () => {
+      // @ts-expect-error BigInt literals are not available when targeting lower than ES2020.
+      const set = new Set([1n, 2n, 3n]);
+      expect(convertToAttr(set)).toEqual({ NS: Array.from(set).map((num) => num.toString()) });
+    });
+
+    it("binary set", () => {
+      const set = new Set([new ArrayBuffer(4), new ArrayBuffer(8), new ArrayBuffer(16)]);
+      expect(convertToAttr(set)).toEqual({ BS: Array.from(set) });
+    });
+
+    it("string set", () => {
+      const set = new Set(["one", "two", "three"]);
+      expect(convertToAttr(set)).toEqual({ SS: Array.from(set) });
+    });
+
+    it("throws error for empty set", () => {
+      expect(() => {
+        convertToAttr(new Set());
+      }).toThrowError(`Please pass a non-empty set`);
+    });
+
+    it("thows error for unallowed set", () => {
+      expect(() => {
+        convertToAttr(new Set([true, false]));
+      }).toThrowError(`Only Number Set (NS), Binary Set (BS) or String Set (SS) are allowed.`);
+    });
+  });
+
+  describe("map", () => {
+    const arr = [...Array(4).keys()];
+    const uint8Arr = new Uint32Array(arr);
+    const biguintArr = new BigUint64Array(arr.map(BigInt));
+    [
+      [
+        { a: null, b: false },
+        { a: { NULL: true }, b: { BOOL: false } },
+      ],
+      [
+        { a: 1.01, b: BigInt(1), c: "one" },
+        { a: { N: "1.01" }, b: { N: "1" }, c: { S: "one" } },
+      ],
+      [
+        { a: uint8Arr, b: biguintArr },
+        { a: { B: uint8Arr }, b: { B: biguintArr } },
+      ],
+    ].forEach(([input, output]) => {
+      it(`testing map: ${input}`, () => {
+        expect(convertToAttr(input)).toEqual({ M: output });
+      });
+    });
+  });
+
+  describe("string", () => {
+    ["", "string", "'single-quote'", '"double-quote"'].forEach((str) => {
+      it(`returns for string: ${str}`, () => {
+        expect(convertToAttr(str)).toEqual({ S: str });
+      });
+    });
+  });
+
+  describe(`unsupported type`, () => {
+    [undefined].forEach((data) => {
```

Comment:
curious -- why throw errors for undefined? 

I would hope we can borrow the rules from JSON serialization... meaning, undefined never makes it to dynamo

Reply:
convertToAttr returns `AttributeValue` to be populated in DynamoDB. If it returns `undefined`, there would be a ValidationException as can be seen in the code below:

<details>
<summary>Code</summary>

```js
const { DynamoDB } = require("@aws-sdk/client-dynamodb");

(async () => {
  const region = "us-west-2";

  const TableName = `test-doc-client`;
  const id = { S: "id1" };
  const undefinedString = { S: undefined };

  const client = new DynamoDB({ region });
  await client.putItem({ TableName, Item: { id, undefinedString } });
  console.log(await client.getItem({ TableName, Key: { id } }));
})();

```

</details>

<details>
<summary>Output</summary>

```console
(node:4711) UnhandledPromiseRejectionWarning: ValidationException: Supplied AttributeValue is empty, must contain exactly one of the supported datatypes
    at deserializeAws_json1_0PutItemCommandError (/Users/trivikr/workspace/ddb-doc-client-test/node_modules/@aws-sdk/client-dynamodb/dist/cjs/protocols/Aws_json1_0.js:1420:41)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:4711) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 6)
(node:4711) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

</details>

convertToAttr ensures that the error is thrown during conversion itself.